File: v2/backend/apps/recipes/services/scale.py
"""Scaling domain service. Next.js does not compute ratio ** 0.7."""

from __future__ import annotations

from dataclasses import dataclass
from decimal import ROUND_HALF_UP, Decimal
from typing import Any

from apps.recipes.constants import UNIT_LABEL_RU

GENTLE_EXPONENT = Decimal("0.7")
# V1 used 1+(ratio-1)*0.5 — never use that here.


class ScaleConflict(Exception):
    """Both servings and anchor_weight were provided — API 400."""


@dataclass(frozen=True)
class ScaleResult:
    enabled: bool
    mode: str  # anchor | servings | off
    ratio: Decimal
    base_anchor: dict[str, Any] | None
    applied: dict[str, Any] | None


def _d(value: Any) -> Decimal:
    if isinstance(value, Decimal):
        return value
    return Decimal(str(value))


def _quantize(value: Decimal, quant: Decimal) -> Decimal:
    return value.quantize(quant, rounding=ROUND_HALF_UP)


def round_scaled(amount: Decimal, unit: str, *, whole: bool = False) -> Decimal:
    """UI rounding from DEFAULTS (port of V1 roundScaled, not the gentle formula)."""
    if whole:
        rounded = _quantize(amount, Decimal("1"))
        return max(Decimal("1"), rounded)

    if unit in {"g", "ml"}:
        if amount >= 200:
            return _quantize(amount / Decimal("10"), Decimal("1")) * Decimal("10")
        if amount >= 50:
            return _quantize(amount / Decimal("5"), Decimal("1")) * Decimal("5")
        return _quantize(amount, Decimal("1"))

    if unit in {"kg", "l"}:
        return _quantize(amount, Decimal("0.01"))

    if unit in {"pcs", "clove", "bunch", "slice"}:
        half = _quantize(amount * 2, Decimal("1")) / Decimal("2")
        return max(Decimal("0.5"), half)

    if unit in {"tsp", "tbsp"}:
        return _quantize(amount * 4, Decimal("1")) / Decimal("4")

    return _quantize(amount, Decimal("0.1"))


def apply_mode(amount: Decimal, ratio: Decimal, scale_mode: str, scalable: bool) -> Decimal:
    if not scalable:
        return amount
    if scale_mode == "manual":
        return amount
    if scale_mode == "gentle":
        return amount * (ratio ** GENTLE_EXPONENT)
    return amount * ratio


def format_display_amount(
    amount: Decimal | None, unit: str, amount_max: Decimal | None = None
) -> str:
    label = UNIT_LABEL_RU.get(unit, unit)
    if unit in {"to_taste", "pinch"} or amount is None:
        return label
    text = _format_number(amount)
    if amount_max is not None:
        text = f"{text}–{_format_number(amount_max)}"
    return f"{text} {label}"


def _format_number(value: Decimal) -> str:
    normalized = value
    if normalized == normalized.to_integral_value():
        return str(int(normalized))
    text = format(normalized, "f")
    if "." in text:
        text = text.rstrip("0").rstrip(".")
    return text.replace(".", ",")


def base_anchor_amount(amount: Decimal, unit: str) -> tuple[Decimal, str] | None:
    """Convert kg/l to g/ml for ratio. Returns (amount, base_unit)."""
    if unit == "kg":
        return amount * Decimal("1000"), "g"
    if unit == "l":
        return amount * Decimal("1000"), "ml"
    if unit in {"g", "ml"}:
        return amount, unit
    return None


def _json_number(value: Decimal) -> int | float:
    if value == value.to_integral_value():
        return int(value)
    return float(value)


def resolve_scale(
    *,
    recipe_scalable: bool,
    recipe_servings: int | None,
    anchor_amount: Decimal | None,
    anchor_unit: str | None,
    servings: Decimal | None,
    anchor_weight: Decimal | None,
    anchor_name: str | None = None,
) -> ScaleResult:
    """Compute ratio from servings XOR anchor_weight.

    No servings and no anchor on the recipe → scaling off; do not invent 4 portions.
    recipe.scalable=false → ignore query, enabled=false.
    """
    if servings is not None and anchor_weight is not None:
        raise ScaleConflict("Нельзя передавать servings и anchor_weight одновременно.")

    off = ScaleResult(
        enabled=False, mode="off", ratio=Decimal("1"), base_anchor=None, applied=None
    )

    if not recipe_scalable:
        return off

    base = None
    if anchor_amount is not None and anchor_unit:
        converted = base_anchor_amount(anchor_amount, anchor_unit)
        if converted:
            base_val, base_unit = converted
            base = {
                "amount": _json_number(base_val),
                "unit": base_unit,
                "name": anchor_name or "",
            }

    has_servings_base = recipe_servings is not None and recipe_servings > 0
    has_anchor_base = base is not None

    if servings is None and anchor_weight is None:
        return off

    if servings is not None:
        if not has_servings_base:
            return off
        ratio = _d(servings) / Decimal(recipe_servings)
        return ScaleResult(
            enabled=True,
            mode="servings",
            ratio=ratio,
            base_anchor=base,
            applied={"servings": _json_number(_d(servings))},
        )

    if not has_anchor_base:
        return off
    ratio = _d(anchor_weight) / _d(base["amount"])
    return ScaleResult(
        enabled=True,
        mode="anchor",
        ratio=ratio,
        base_anchor=base,
        applied={"anchor_weight": _json_number(_d(anchor_weight))},
    )


def scale_line(
    *,
    amount: Decimal | None,
    amount_max: Decimal | None,
    unit: str,
    scale_mode: str,
    scalable: bool,
    ratio: Decimal,
    scaling_enabled: bool,
) -> tuple[Decimal | None, Decimal | None]:
    """Return scaled (and rounded) amount / amount_max. Timers/temps are never scaled."""
    if amount is None:
        return None, None
    if not scaling_enabled:
        return amount, amount_max

    whole = scale_mode == "whole" and scalable
    raw = apply_mode(amount, ratio, scale_mode, scalable)
    out = round_scaled(raw, unit, whole=whole)
    out_max = None
    if amount_max is not None:
        raw_max = apply_mode(amount_max, ratio, scale_mode, scalable)
        out_max = round_scaled(raw_max, unit, whole=whole)
    return out, out_max

File: v2/frontend/lib/scale.ts
import type { RecipeIngredient } from '@/lib/types';

const GENTLE_EXPONENT = 0.7;

const UNIT_LABEL: Record<string, string> = {
  g: 'г',
  kg: 'кг',
  ml: 'мл',
  l: 'л',
  pcs: 'шт',
  tsp: 'ч. л.',
  tbsp: 'ст. л.',
  pinch: 'щепотка',
  clove: 'зубчик',
  bunch: 'пучок',
  slice: 'ломтик',
  to_taste: 'по вкусу',
};

function roundHalfUp(value: number, quantum: number): number {
  const n = value / quantum;
  const sign = n < 0 ? -1 : 1;
  const abs = Math.abs(n);
  const whole = Math.floor(abs + 1e-12);
  const frac = abs - whole;
  const rounded = frac >= 0.5 - 1e-12 ? whole + 1 : whole;
  return sign * rounded * quantum;
}

export function applyMode(
  amount: number,
  ratio: number,
  scaleMode: string,
  scalable: boolean,
): number {
  if (!scalable) return amount;
  if (scaleMode === 'manual') return amount;
  if (scaleMode === 'gentle') return amount * ratio ** GENTLE_EXPONENT;
  return amount * ratio;
}

export function roundScaled(amount: number, unit: string, whole = false): number {
  if (whole) return Math.max(1, roundHalfUp(amount, 1));
  if (unit === 'g' || unit === 'ml') {
    if (amount >= 200) return roundHalfUp(amount / 10, 1) * 10;
    if (amount >= 50) return roundHalfUp(amount / 5, 1) * 5;
    return roundHalfUp(amount, 1);
  }
  if (unit === 'kg' || unit === 'l') return roundHalfUp(amount, 0.01);
  if (unit === 'pcs' || unit === 'clove' || unit === 'bunch' || unit === 'slice') {
    return Math.max(0.5, roundHalfUp(amount, 0.5));
  }
  if (unit === 'tsp' || unit === 'tbsp') return roundHalfUp(amount, 0.25);
  return roundHalfUp(amount, 0.1);
}

function formatNumber(value: number): string {
  if (!Number.isFinite(value)) return '';
  if (Math.abs(value - Math.round(value)) < 1e-9) return String(Math.round(value));
  const text = value.toFixed(6).replace(/\.?0+$/, '');
  return text.replace('.', ',');
}

const SPOON_ML: Record<string, number> = { tsp: 5, tbsp: 15 };

function spoonMl(amount: number, unit: string): string {
  return formatNumber(roundHalfUp(amount * SPOON_ML[unit], 1));
}

export function formatDisplayAmount(
  amount: number | null,
  unit: string,
  amountMax: number | null = null,
): string {
  const label = UNIT_LABEL[unit] ?? unit;
  if (unit === 'to_taste' || unit === 'pinch' || amount == null) return label;
  let text = formatNumber(amount);
  if (amountMax != null) text = `${text}–${formatNumber(amountMax)}`;
  let shown = `${text} ${label}`;
  if (unit in SPOON_ML) {
    let hint = spoonMl(amount, unit);
    if (amountMax != null) hint = `${hint}–${spoonMl(amountMax, unit)}`;
    shown = `${shown} (~${hint} мл)`;
  }
  return shown;
}

export function scaleLine(
  line: RecipeIngredient,
  ratio: number,
): RecipeIngredient {
  if (line.amount == null) {
    return { ...line, display_amount: formatDisplayAmount(null, line.unit, line.amount_max) };
  }
  const whole = line.scale_mode === 'whole' && line.scalable;
  const raw = applyMode(line.amount, ratio, line.scale_mode, line.scalable);
  const amount = roundScaled(raw, line.unit, whole);
  let amountMax: number | null = null;
  if (line.amount_max != null) {
    amountMax = roundScaled(
      applyMode(line.amount_max, ratio, line.scale_mode, line.scalable),
      line.unit,
      whole,
    );
  }
  return {
    ...line,
    amount,
    amount_max: amountMax,
    display_amount: formatDisplayAmount(amount, line.unit, amountMax),
  };
}

export function scaleIngredients(
  lines: RecipeIngredient[],
  ratio: number,
): RecipeIngredient[] {
  if (!Number.isFinite(ratio) || Math.abs(ratio - 1) < 0.0001) return lines;
  return lines.map((line) => scaleLine(line, ratio));
}
